File: lib/rust/src/redis.rs
use std::sync::Arc;
use std::time::Duration;
use redis::{
    aio::ConnectionManager,
    AsyncCommands, Client, RedisError,
};
use futures_util::StreamExt;
use ipnetwork::IpNetwork;
use serde::{Deserialize, Serialize};
use tokio::sync::{mpsc, Mutex, broadcast};
use tokio::task::JoinHandle;
use tracing::{info, error, debug};

use crate::traits::RadixEngine;
use crate::types::Metadata;

// Custom error type
#[derive(Debug, thiserror::Error)]
pub enum RedisPubSubError {
    #[error("Redis error: {0}")]
    Redis(#[from] RedisError),
    #[error("Channel send error")]
    SendError,
    #[error("Channel receive error")]
    RecvError,
}

type Result<T> = std::result::Result<T, RedisPubSubError>;

// Configuration for Redis connection
#[derive(Debug, Clone)]
pub struct RedisConfig {
    pub url: String,
    pub pool_size: usize,
    pub connect_timeout: Duration,
    pub max_retries: usize,
}

impl Default for RedisConfig {
    fn default() -> Self {
        Self {
            url: "redis://127.0.0.1:6379".to_string(),
            pool_size: 10,
            connect_timeout: Duration::from_secs(5),
            max_retries: 3,
        }
    }
}

// Thread-safe Redis client wrapper
#[derive(Clone)]
pub struct RedisClient {
    inner: Arc<RedisClientInner>,
}

struct RedisClientInner {
    client: Client,
    connection_manager: Mutex<ConnectionManager>,
    config: RedisConfig,
    pubsub_sender: broadcast::Sender<PubSubMessage>,
    shutdown_tx: broadcast::Sender<()>,
}

#[derive(Debug, Clone)]
pub struct PubSubMessage {
    pub channel: String,
    pub payload: String,
    pub pattern: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "op", rename_all = "snake_case")]
pub enum RedisCacheUpdate {
    Insert {
        prefix: IpNetwork,
        metadata: Metadata,
    },
    Remove {
        prefix: IpNetwork,
    },
    Clear,
}

impl RedisClient {
    /// Create a new Redis client with connection pooling
    pub async fn new(config: RedisConfig) -> Result<Self> {
        let client = Client::open(config.url.clone())?;
        
        // Create connection manager with retry logic
        let connection_manager = ConnectionManager::new(client.clone())
            .await
            .map_err(|e| RedisPubSubError::Redis(e))?;

        let (pubsub_tx, _) = broadcast::channel(100);
        let (shutdown_tx, _) = broadcast::channel(1);

        let inner = RedisClientInner {
            client,
            connection_manager: Mutex::new(connection_manager),
            config,
            pubsub_sender: pubsub_tx,
            shutdown_tx,
        };

        Ok(Self {
            inner: Arc::new(inner),
        })
    }

    /// Get a connection from the pool
    async fn get_connection(&self) -> Result<tokio::sync::MutexGuard<'_, ConnectionManager>> {
        Ok(self.inner.connection_manager.lock().await)
    }

    /// Get a synchronous connection
    pub fn get_sync_connection(&self) -> Result<redis::Connection> {
        self.inner.client.get_connection().map_err(RedisPubSubError::Redis)
    }

        /// Publish a message to a channel
    pub async fn publish(&self, channel: &str, message: &str) -> Result<()> {
        let mut conn = self.get_connection().await?;
        let _: usize = conn.publish(channel, message).await?;
        debug!("Published to {}: {}", channel, message);
        Ok(())
    }

    /// Publish a JSON message
    pub async fn publish_json<T: serde::Serialize>(&self, channel: &str, data: &T) -> Result<()> {
        let json = serde_json::to_string(data)
            .map_err(|e| RedisPubSubError::Redis(RedisError::from((
                redis::ErrorKind::TypeError,
                "Serialization error",
                e.to_string(),
            ))))?;
        self.publish(channel, &json).await
    }

    /// Subscribe to a channel and process messages with a callback
    pub async fn subscribe<F, Fut>(&self, channel: &str, mut callback: F) -> Result<JoinHandle<()>>
    where
        F: FnMut(PubSubMessage) -> Fut + Send + 'static,
        Fut: std::future::Future<Output = ()> + Send,
    {
        let client = self.inner.client.clone();
        let channel_name = channel.to_string();
        let shutdown_rx = self.inner.shutdown_tx.subscribe();
        
        let handle = tokio::spawn(async move {
            let mut shutdown = shutdown_rx;
            let mut pubsub = match client.get_async_pubsub().await {
                Ok(pubsub) => pubsub,
                Err(error) => {
                    error!("Failed to create Redis Pub/Sub connection: {}", error);
                    return;
                }
            };
            if let Err(error) = pubsub.subscribe(&channel_name).await {
                error!("Failed to subscribe to {}: {}", channel_name, error);
                return;
            }
            let mut stream = pubsub.on_message();

            loop {
                tokio::select! {
                    msg_result = stream.next() => {
                        match msg_result {
                            Some(msg) => {
                                let payload: String = msg.get_payload().unwrap_or_default();
                                let channel_name = msg.get_channel_name().to_string();
                                
                                let pubsub_msg = PubSubMessage {
                                    channel: channel_name,
                                    payload,
                                    pattern: None,
                                };
                                
                                callback(pubsub_msg).await;
                            }
                            None => {
                                debug!("PubSub stream ended for channel {}", channel_name);
                                break;
                            }
                        }
                    }
                    _ = shutdown.recv() => {
                        info!("Shutting down subscription for channel {}", channel_name);
                        break;
                    }
                }
            }
        });

        Ok(handle)
    }

     /// Subscribe to a channel and send messages to an mpsc channel
    pub async fn subscribe_to_channel(
        &self,
        channel: &str,
    ) -> Result<(mpsc::Receiver<PubSubMessage>, JoinHandle<()>)> {
        let (tx, rx) = mpsc::channel(100);
        let tx_clone = tx.clone();

        let client = self.inner.client.clone();
        let channel_name = channel.to_string();
        let shutdown_rx = self.inner.shutdown_tx.subscribe();

        let handle = tokio::spawn(async move {
            let mut shutdown = shutdown_rx;
            let mut pubsub = match client.get_async_pubsub().await {
                Ok(pubsub) => pubsub,
                Err(error) => {
                    error!("Failed to create Redis Pub/Sub connection: {}", error);
                    return;
                }
            };
            if let Err(error) = pubsub.subscribe(&channel_name).await {
                error!("Failed to subscribe to {}: {}", channel_name, error);
                return;
            }
            let mut stream = pubsub.on_message();

            loop {
                tokio::select! {
                    msg_result = stream.next() => {
                        match msg_result {
                            Some(msg) => {
                                let payload: String = msg.get_payload().unwrap_or_default();
                                let channel_name = msg.get_channel_name().to_string();
                                
                                let pubsub_msg = PubSubMessage {
                                    channel: channel_name,
                                    payload,
                                    pattern: None,
                                };
                                
                                if let Err(e) = tx_clone.send(pubsub_msg).await {
                                    error!("Failed to send message to channel: {}", e);
                                    break;
                                }
                            }
                            None => {
                                debug!("PubSub stream ended for channel {}", channel_name);
                                break;
                            }
                        }
                    }
                    _ = shutdown.recv() => {
                        info!("Shutting down subscription for channel {}", channel_name);
                        break;
                    }
                }
            }
        });

        Ok((rx, handle))
    }

       /// Subscribe to a pattern (e.g., "news.*")
    pub async fn psubscribe(
        &self,
        pattern: &str,
    ) -> Result<(mpsc::Receiver<PubSubMessage>, JoinHandle<()>)> {
        let (tx, rx) = mpsc::channel(100);
        let tx_clone = tx.clone();

        let client = self.inner.client.clone();
        let shutdown_rx = self.inner.shutdown_tx.subscribe();
        let pattern_str = pattern.to_string();

        let handle = tokio::spawn(async move {
            let mut shutdown = shutdown_rx;
            let mut pubsub = match client.get_async_pubsub().await {
                Ok(pubsub) => pubsub,
                Err(error) => {
                    error!("Failed to create Redis Pub/Sub connection: {}", error);
                    return;
                }
            };
            if let Err(error) = pubsub.psubscribe(&pattern_str).await {
                error!("Failed to subscribe to pattern {}: {}", pattern_str, error);
                return;
            }
            let mut stream = pubsub.on_message();

            loop {
                tokio::select! {
                    msg_result = stream.next() => {
                        match msg_result {
                            Some(msg) => {
                                let payload: String = msg.get_payload().unwrap_or_default();
                                let channel_name = msg.get_channel_name().to_string();
                                let pattern = msg.get_pattern::<String>().ok();
                                
                                let pubsub_msg = PubSubMessage {
                                    channel: channel_name,
                                    payload,
                                    pattern,
                                };
                                
                                if let Err(e) = tx_clone.send(pubsub_msg).await {
                                    error!("Failed to send message to channel: {}", e);
                                    break;
                                }
                            }
                            None => {
                                debug!("PubSub stream ended for pattern {}", pattern_str);
                                break;
                            }
                        }
                    }
                    _ = shutdown.recv() => {
                        info!("Shutting down subscription for pattern {}", pattern_str);
                        break;
                    }
                }
            }
        });

        Ok((rx, handle))
    }

      /// Broadcast to all subscribers
    pub fn broadcast(&self) -> broadcast::Sender<PubSubMessage> {
        self.inner.pubsub_sender.clone()
    }

    /// Subscribe to broadcast messages
    pub fn subscribe_broadcast(&self) -> broadcast::Receiver<PubSubMessage> {
        self.inner.pubsub_sender.subscribe()
    }

    pub fn config(&self) -> &RedisConfig {
        &self.inner.config
    }

      /// Shutdown all subscriptions
    pub async fn shutdown(&self) -> Result<()> {
        let _ = self.inner.shutdown_tx.send(());
        Ok(())
    }

    /// Set a key-value pair
    pub async fn set(&self, key: &str, value: &str) -> Result<()> {
        let mut conn = self.get_connection().await?;
        let _: () = conn.set(key, value).await?;
        Ok(())
    }

    /// Get a value by key
    pub async fn get(&self, key: &str) -> Result<Option<String>> {
        let mut conn = self.get_connection().await?;
        let result: Option<String> = conn.get(key).await?;
        Ok(result)
    }

    /// Synchronous Set
    pub fn set_sync(&self, key: &str, value: &str) -> Result<()> {
        let mut conn = self.get_sync_connection()?;
        redis::cmd("SET").arg(key).arg(value).query(&mut conn)?;
        Ok(())
    }

    /// Synchronous Get
    pub fn get_sync(&self, key: &str) -> Result<Option<String>> {
        let mut conn = self.get_sync_connection()?;
        let result: Option<String> = redis::cmd("GET").arg(key).query(&mut conn)?;
        Ok(result)
    }

    /// Synchronous HGetAll for boot-loading prefixes
    pub fn hgetall_sync(&self, key: &str) -> Result<std::collections::HashMap<String, String>> {
        let mut conn = self.get_sync_connection()?;
        let result: std::collections::HashMap<String, String> = redis::cmd("HGETALL").arg(key).query(&mut conn)?;
        Ok(result)
    }
    
    /// Synchronous HSet
    pub fn hset_sync(&self, key: &str, field: &str, value: &str) -> Result<()> {
        let mut conn = self.get_sync_connection()?;
        redis::cmd("HSET").arg(key).arg(field).arg(value).query(&mut conn)?;
        Ok(())
    }

    /// Synchronous HDel
    pub fn hdel_sync(&self, key: &str, field: &str) -> Result<()> {
        let mut conn = self.get_sync_connection()?;
        redis::cmd("HDEL").arg(key).arg(field).query(&mut conn)?;
        Ok(())
    }

    pub async fn publish_insert(
        &self,
        channel: &str,
        prefix: IpNetwork,
        metadata: Metadata,
    ) -> Result<()> {
        self.publish_json(
            channel,
            &RedisCacheUpdate::Insert { prefix, metadata },
        )
        .await
    }

    pub async fn publish_remove(&self, channel: &str, prefix: IpNetwork) -> Result<()> {
        self.publish_json(channel, &RedisCacheUpdate::Remove { prefix }).await
    }

    pub async fn publish_clear(&self, channel: &str) -> Result<()> {
        self.publish_json(channel, &RedisCacheUpdate::Clear).await
    }

    pub async fn subscribe_engine_updates(
        &self,
        channel: &str,
        engine: Arc<dyn RadixEngine>,
    ) -> Result<JoinHandle<()>> {
        self.subscribe(channel, move |message| {
            let engine = engine.clone();
            async move {
                match serde_json::from_str::<RedisCacheUpdate>(&message.payload) {
                    Ok(RedisCacheUpdate::Insert { prefix, metadata }) => {
                        if let Err(error) = engine.insert(prefix, metadata) {
                            error!("Failed to apply Redis insert update: {}", error);
                        }
                    }
                    Ok(RedisCacheUpdate::Remove { prefix }) => {
                        engine.remove(&prefix);
                    }
                    Ok(RedisCacheUpdate::Clear) => {
                        engine.clear();
                    }
                    Err(error) => {
                        error!("Failed to decode Redis cache update: {}", error);
                    }
                }
            }
        })
        .await
    }
}

File: lib/go/redis.go
package radixip

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"sync"
	"time"

	"github.com/redis/go-redis/v9"
)

// Custom error types
var (
	ErrSendError = errors.New("channel send error")
	ErrRecvError = errors.New("channel receive error")
)

// Configuration for Redis connection
type RedisConfig struct {
	URL            string
	PoolSize       int
	ConnectTimeout time.Duration
	MaxRetries     int
}

func DefaultRedisConfig() RedisConfig {
	return RedisConfig{
		URL:            "redis://127.0.0.1:6379",
		PoolSize:       10,
		ConnectTimeout: 5 * time.Second,
		MaxRetries:     3,
	}
}

// PubSubMessage represents a message from Redis pub/sub
type PubSubMessage struct {
	Channel string
	Payload string
	Pattern *string
}

// RedisCacheUpdate represents cache update operations
type RedisCacheUpdate struct {
	Op       string     `json:"op"`
	Prefix   *IpNetwork `json:"prefix,omitempty"`
	Metadata *Metadata  `json:"metadata,omitempty"`
}

// Cache update operation constants
const (
	OpInsert = "insert"
	OpRemove = "remove"
	OpClear  = "clear"
)

// RedisClient wraps the go-redis client
type RedisClient struct {
	inner *RedisClientInner
}

type RedisClientInner struct {
	client        *redis.Client
	config        RedisConfig
	pubsubSender  chan PubSubMessage
	shutdownCh    chan struct{}
	subscriptions sync.WaitGroup
	mu            sync.RWMutex
	subscribers   map[string][]chan PubSubMessage
}

// NewRedisClient creates a new Redis client
func NewRedisClient(config RedisConfig) (*RedisClient, error) {
	rdb := redis.NewClient(&redis.Options{
		Addr:         config.URL,
		PoolSize:     config.PoolSize,
		DialTimeout:  config.ConnectTimeout,
		MaxRetries:   config.MaxRetries,
		ReadTimeout:  10 * time.Second,
		WriteTimeout: 10 * time.Second,
	})

	// Test connection
	ctx, cancel := context.WithTimeout(context.Background(), config.ConnectTimeout)
	defer cancel()

	if err := rdb.Ping(ctx).Err(); err != nil {
		return nil, fmt.Errorf("failed to connect to Redis: %w", err)
	}

	inner := &RedisClientInner{
		client:       rdb,
		config:       config,
		pubsubSender: make(chan PubSubMessage, 100),
		shutdownCh:   make(chan struct{}),
		subscribers:  make(map[string][]chan PubSubMessage),
	}

	return &RedisClient{inner: inner}, nil
}

// Publish publishes a message to a channel
func (r *RedisClient) Publish(ctx context.Context, channel string, message string) error {
	err := r.inner.client.Publish(ctx, channel, message).Err()
	if err != nil {
		return fmt.Errorf("failed to publish message: %w", err)
	}
	return nil
}

// PublishJSON publishes a JSON-serialized message to a channel
func (r *RedisClient) PublishJSON(ctx context.Context, channel string, data interface{}) error {
	jsonData, err := json.Marshal(data)
	if err != nil {
		return fmt.Errorf("failed to serialize message: %w", err)
	}
	return r.Publish(ctx, channel, string(jsonData))
}

// Subscribe subscribes to a channel and processes messages with a callback
func (r *RedisClient) Subscribe(ctx context.Context, channel string, callback func(PubSubMessage)) error {
	pubsub := r.inner.client.Subscribe(ctx, channel)
	defer pubsub.Close()

	ch := pubsub.Channel()

	r.inner.mu.Lock()
	if r.inner.subscribers[channel] == nil {
		r.inner.subscribers[channel] = make([]chan PubSubMessage, 0)
	}
	//100 messages at a time
	msgCh := make(chan PubSubMessage, 100)
	r.inner.subscribers[channel] = append(r.inner.subscribers[channel], msgCh)
	r.inner.mu.Unlock()

	r.inner.subscriptions.Add(1)
	defer r.inner.subscriptions.Done()

	for {
		select {
		case msg := <-ch:
			pubsubMsg := PubSubMessage{
				Channel: msg.Channel,
				Payload: msg.Payload,
				Pattern: msg.Pattern,
			}

			// Send to internal subscribers
			r.inner.mu.RLock()
			for _, ch := range r.inner.subscribers[channel] {
				select {
				case ch <- pubsubMsg:
				default:
					// Channel full, skip
				}
			}
			r.inner.mu.RUnlock()

			// Call the callback
			callback(pubsubMsg)

		case <-r.inner.shutdownCh:
			return nil
		case <-ctx.Done():
			return ctx.Err()
		}
	}
}

// SubscribeToChannel subscribes to a channel and returns a channel of messages
func (r *RedisClient) SubscribeToChannel(ctx context.Context, channel string) (<-chan PubSubMessage, error) {
	pubsub := r.inner.client.Subscribe(ctx, channel)
	ch := pubsub.Channel()

	msgCh := make(chan PubSubMessage, 100)

	r.inner.subscriptions.Add(1)
	go func() {
		defer r.inner.subscriptions.Done()
		defer close(msgCh)

		for {
			select {
			case msg := <-ch:
				msgCh <- PubSubMessage{
					Channel: msg.Channel,
					Payload: msg.Payload,
					Pattern: msg.Pattern,
				}
			case <-r.inner.shutdownCh:
				return
			case <-ctx.Done():
				return
			}
		}
	}()

	return msgCh, nil
}

// PSubscribe subscribes to a pattern and returns a channel of messages
func (r *RedisClient) PSubscribe(ctx context.Context, pattern string) (<-chan PubSubMessage, error) {
	pubsub := r.inner.client.PSubscribe(ctx, pattern)
	ch := pubsub.Channel()

	msgCh := make(chan PubSubMessage, 100)

	r.inner.subscriptions.Add(1)
	go func() {
		defer r.inner.subscriptions.Done()
		defer close(msgCh)

		for {
			select {
			case msg := <-ch:
				msgCh <- PubSubMessage{
					Channel: msg.Channel,
					Payload: msg.Payload,
					Pattern: msg.Pattern,
				}
			case <-r.inner.shutdownCh:
				return
			case <-ctx.Done():
				return
			}
		}
	}()

	return msgCh, nil
}

// Broadcast sends a message to all internal subscribers
func (r *RedisClient) Broadcast(msg PubSubMessage) {
	select {
	case r.inner.pubsubSender <- msg:
	default:
		// Channel full, skip
	}
}

// SubscribeBroadcast returns a channel for broadcast messages
func (r *RedisClient) SubscribeBroadcast() <-chan PubSubMessage {
	return r.inner.pubsubSender
}

// GetConfig returns the Redis configuration
func (r *RedisClient) GetConfig() RedisConfig {
	return r.inner.config
}

// Shutdown shuts down all subscriptions
func (r *RedisClient) Shutdown() {
	close(r.inner.shutdownCh)
	r.inner.subscriptions.Wait()
}

// Set sets a key-value pair
func (r *RedisClient) Set(ctx context.Context, key string, value string) error {
	return r.inner.client.Set(ctx, key, value, 0).Err()
}

// Get gets a value by key
func (r *RedisClient) Get(ctx context.Context, key string) (string, error) {
	return r.inner.client.Get(ctx, key).Result()
}

// HSet sets a field in a hash
func (r *RedisClient) HSet(ctx context.Context, key string, field string, value string) error {
	return r.inner.client.HSet(ctx, key, field, value).Err()
}

// HGetAll gets all fields in a hash
func (r *RedisClient) HGetAll(ctx context.Context, key string) (map[string]string, error) {
	return r.inner.client.HGetAll(ctx, key).Result()
}

// HDel deletes a field in a hash
func (r *RedisClient) HDel(ctx context.Context, key string, field string) error {
	return r.inner.client.HDel(ctx, key, field).Err()
}

// PublishInsert publishes an insert operation
func (r *RedisClient) PublishInsert(ctx context.Context, channel string, prefix IpNetwork, metadata Metadata) error {
	update := RedisCacheUpdate{
		Op:       OpInsert,
		Prefix:   &prefix,
		Metadata: &metadata,
	}
	return r.PublishJSON(ctx, channel, update)
}

// PublishRemove publishes a remove operation
func (r *RedisClient) PublishRemove(ctx context.Context, channel string, prefix IpNetwork) error {
	update := RedisCacheUpdate{
		Op:     OpRemove,
		Prefix: &prefix,
	}
	return r.PublishJSON(ctx, channel, update)
}

// PublishClear publishes a clear operation
func (r *RedisClient) PublishClear(ctx context.Context, channel string) error {
	update := RedisCacheUpdate{
		Op: OpClear,
	}
	return r.PublishJSON(ctx, channel, update)
}

// SubscribeEngineUpdates subscribes to engine updates and applies them
func (r *RedisClient) SubscribeEngineUpdates(ctx context.Context, channel string, engine RadixEngine) error {
	return r.Subscribe(ctx, channel, func(msg PubSubMessage) {
		var update RedisCacheUpdate
		if err := json.Unmarshal([]byte(msg.Payload), &update); err != nil {
			// Log error
			return
		}

		switch update.Op {
		case OpInsert:
			if update.Prefix != nil && update.Metadata != nil {
				if err := engine.Insert(*update.Prefix, *update.Metadata); err != nil {
					// Log error
				}
			}
		case OpRemove:
			if update.Prefix != nil {
				engine.Remove(*update.Prefix)
			}
		case OpClear:
			engine.Clear()
		}
	})
}
